Enhancement of markers

nem1234 opened the pull request:
Enhancement of markers to support Digital Micrograph Annotations (PR #1491)

### Description of the change
1. add filled rectangle / text box
1. new marker : arrow, ellipse

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np
mat = np.arange(10000).reshape(100,100)
data = hs.signals.Signal2D(mat)
m = hs.plot.markers.rectangle(50,50,60,60,edgecolor="red",facecolor="green",fill=True,zorder=3)
data.add_marker(m)
```

nem1234 committed:
Enhancement of markers
1. zorder property
2. filled rectangle / text box
3. new marker : arrow, ellipse

nem1234 committed:
Add documents
Add aliases in hyperspy.plot.markers
Change parameters in marker.ellipse

nem1234 committed:
bugfix rectangle parameters

nem1234 committed:
modify test to accept current marker setting

nem1234 committed:
fix test for arrow

nem1234 committed:
fix typo

nem1234 committed:
Add change log

nem1234 committed:
Update document and doc strings

nem1234 committed:
revise copyright year

nem1234 committed:
Revise test code to add 'arrow' and 'ellipse' into directory.

nem1234 committed:
Reformat source style.
Remove 'color' property for arrow and ellipse
Add test code for rectangle

nem1234 committed:
Add test for marker parameters

nem1234 committed:
revise test for ellipse parameters

nem1234 committed:
Fix typo
Add repr linewidth parameter for rectangle

nem1234 committed:
Revise test code and repr string for point

nem1234 committed:
Arrow: change from ax.annotation to patches.FancyArrowPatch
Text: revert to ax.Text

nem1234 committed:
Add test for iterable parameters of arrow and ellipse

nem1234 committed:
arrow : set default parameter of shrinkA,B to 0
ellipse : bug fix of auto update
add test code of iterativ parameters for arrow and ellipse

nem1234 committed:
Modify test code for properties in iterated markers

nem1234 committed:
Add import for logger

nem1234 committed:
Add test, modify parameters for arrow and ellipse

nem1234 committed:
Add code for older matplotlib

jlaehne commented:
The failing tests are now fixed, maybe you can merge in Release_next_minor

jlaehne commented:
Great that someone picks up on this.

There are few codecov warnings when viewing the file changes, maybe you can use them to further improve coverage.

jlaehne commented on doc/user_guide/visualisation.rst:
```diff
@@ -1332,3 +1332,28 @@ Permanent markers are stored in the HDF5 file if the signal is saved:
+Signal plane has zorder=0, and the marker that have largest number
+become a top most marker.
+  
+Currently these markers are provided in hyperspy. The 'color' property in
```
HyperSpy

nem1234 replied:
Fixed.

jlaehne commented on doc/user_guide/visualisation.rst:
```diff
@@ -1332,3 +1332,28 @@ Permanent markers are stored in the HDF5 file if the signal is saved:
+Currently these markers are provided in hyperspy. The 'color' property in
+rectangle marker is an alias of 'edgecolor' for backward compatibility.
+The other parameters are used as in the parameters in matplotlib patches.
+The parameters after @@ are typical optional parameters.
```
Please see how optional parameters are given elsewhere instead of introducing an individual style.

nem1234 replied:
options for matplotlib objects are replaced with **kwargs

jlaehne commented on hyperspy/drawing/_markers/arrow.py:
```diff
@@ -0,0 +1,110 @@
+
+class Arrow(MarkerBase):
+
+    """Rectangle marker that can be added to the signal figure
```
rectangle -> arrow
Here and in the following

nem1234 replied:
Fixed.

jlaehne commented on hyperspy/drawing/_markers/ellipse.py:
```diff
@@ -0,0 +1,109 @@
+        The width (diameter) of the ellipse in x.
+        see x arguments
+    height : array or float
+        The hight (diameter) of the ellipse in y.
```
height

nem1234 replied:
Fixed.

nem1234 committed:
Update document/docstrings for markers
Fix typo in horizontal_line_segment()

nem1234 commented:
Thank you for your advice.
The document 'visualization.rst' is revised.
I also tried to revise the URL link in visualization.rst, but I couldn't find the new URL of EDS tutorial. 
Where is it?

nem1234 committed:
Escape ** in visualize.rst

jlaehne commented on doc/user_guide/visualisation.rst:
```diff
@@ -445,10 +445,10 @@ Data files used in the following examples can be downloaded using
     `EDS tutorials <http://nbviewer.ipython.org/github/hyperspy/hyperspy-demos/blob/master/electron_microscopy/EDS/>`_ .
```
The new nbviewer link seems to be:
https://nbviewer.org/github/hyperspy/hyperspy-demos/tree/main/electron_microscopy/EDS/

Can you check if that works? The `master` branch was renamed to `main` a while ago. I think that is the problem

jlaehne replied:
Can you still correct this link?

nem1234 replied:
Ok, I'll replace the link

jlaehne replied:
There is three more occurences of the same link, can you also correct those.

ericpre commented on hyperspy/drawing/_markers/arrow.py:
```diff
@@ -0,0 +1,110 @@
+        self.name = 'arrow'
+
+    def __repr__(self):
+        string = "<marker.{}, {} (x1={},x2={},y1={},y2={},edgecolor={},facecolor={},linewidth={},arrowstyle={},mutation_scale={},zorder={})>".format(
```
When we display the representation of an python, we are interested in the main information about this object, having too much information is a bit distractive. So I don't know if this is useful to have all this information to represent the object.
If one want more information, this can be easily accessed through the `markers_properties` attribute.

nem1234 replied:
repr strings in markers are simplified.

ericpre commented on doc/user_guide/visualisation.rst:
```diff
@@ -1332,3 +1332,25 @@ Permanent markers are stored in the HDF5 file if the signal is saved:
+The other parameters are used as in the parameters in matplotlib objects.
+The optional parameters (\*\*kwargs, keyword arguments) can be used for extra parameters used in matplotlib.
+
+.. code-block:: python
```
For some reason, the code block is not displayed in the user guide.

nem1234 replied:
add blank line after code-block statement.
I hope this resolve the problem.

jlaehne replied:
can you use also `>>>` as in the code block above.

Also, the links do not work in code-block. 
- Either you just put the URLs (and maybe the comments on an extra line as it is otherwise so long).
- Or you put all the references in an extra block after the code-block. Then linking will also work. You can even use [intersphinx linking](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html>) -- should be e.g. ``` :external:py:class:`matplotlib.patches.FancyArrowPatch` ```

ericpre commented on doc/user_guide/visualisation.rst:
```diff
@@ -1332,3 +1332,25 @@ Permanent markers are stored in the HDF5 file if the signal is saved:
     >>> s1.metadata.Markers
     └── point = <hyperspy.drawing._markers.point.Point object at 0x7efcfadb06d8>
+
+Markers have 'zorder' property to support ordering between markers.
```
If you want to add some code highlighting, as it is usually done:
```suggestion
Markers have ``'zorder'`` property to support ordering between markers.
```

nem1234 replied:
fixed. "'"'s are replaced with "``".

ericpre reviewed (commented):
The failure on azure pipelines are related to the image comparison failure. For image comparison testing, we need to use specific version of matplotlib, when generating/updating the baseline images, make sure to have 
https://github.com/hyperspy/hyperspy/blob/2856c925c7e89c7d30b4c51e00b40207a8851d3d/conda_environment_dev.yml#L6-L7
and see for https://hyperspy--2871.org.readthedocs.build/en/2871/dev_guide/testing.html#plot-testing for more information.

Here are failures:
```python
_________________________ test_plot_rectangle_markers __________________________
[gw0] linux -- Python 3.8.12 /home/vsts/work/1/mambaforge/envs/test_env/bin/python
Image file not found for comparison test in: 
	/home/vsts/work/1/s/hyperspy/tests/drawing/plot_markers
(This is expected for new tests.)
Generated Image: 
	/home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_plot_rectangle_markers/result.png
```
```python
_____________________________ test_markers_zorder ______________________________
[gw0] linux -- Python 3.8.12 /home/vsts/work/1/mambaforge/envs/test_env/bin/python
Error: Image files did not match.
  RMS Value: 16.185499749599686
  Expected:  
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder/baseline.png
  Actual:    
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder/result.png
  Difference:
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder/result-failed-diff.png
  Tolerance: 
    2.0

hyperspy/tests/drawing/test_plot_markers.py RRR.                         [ 17%]
hyperspy/tests/drawing/test_plot_mva.py F

_____________________________ test_markers_zorder2 _____________________________
[gw0] linux -- Python 3.8.12 /home/vsts/work/1/mambaforge/envs/test_env/bin/python
Error: Image files did not match.
  RMS Value: 16.185499749599686
  Expected:  
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder2/baseline.png
  Actual:    
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder2/result.png
  Difference:
    /home/vsts/work/1/s/result_images/hyperspy.tests.drawing.test_plot_markers.test_markers_zorder2/result-failed-diff.png
  Tolerance: 
    2.0

```

nem1234 commented:
> The failure on azure pipelines are related to the image comparison failure. For image comparison testing,
Thank you.
But, when matplotlib-3.1.3 is used for base image generation, what happens in Tests / ubuntu-py3.9 ?
Under Tests/ubuntu-py3.9, matplotlib 3.4.3 is used for image generation.
I don't know how to pass test for both versions of matplotlib.
Remove image compare from tests, generating images for every versions of matplotlib, create custom text drawing code, ....

Any suggestions?

nem1234 commented:
> The failure on azure pipelines are related to the image comparison failure. For image comparison testing, we need to use specific version of matplotlib,
I tried to re-generate test images, but the text drawings were little bit different to the original ones, even if matplotlib=3.1.3 was used.
May be due to the font setting, but it is difficult to find correct settings.
If it is difficult to correct it, I'll remove text drawing from test image.

nem1234 committed:
Simplify repr strings
brush up document

ericpre commented:
> > The failure on azure pipelines are related to the image comparison failure. For image comparison testing, we need to use specific version of matplotlib,
> > I tried to re-generate test images, but the text drawings were little bit different to the original ones, even if matplotlib=3.1.3 was used.
> > May be due to the font setting, but it is difficult to find correct settings.
> > If it is difficult to correct it, I'll remove text drawing from test image.

What is your version freetype? Depending on how you installed matplotlib, the freetype version may be included with matplotlib - I think this is the case for pypi packages but I am not sure. If you have use matplotlib package from conda-forge, you can specify the freetype version used by matplotlib by changing the freetype version using conda.
Alternatively, there is an option to remove text, see [pytest-mpl documentation](https://github.com/matplotlib/pytest-mpl#removing-text).

If this is a pain or not working, I can update the baseline images for this pull request.

nem1234 committed:
Add support matplotlib-3.1.x for ellipse marker

nem1234 commented:
I've tried to build hyperspy environment with freetype==2.9 and matplotlib==3.1.3 using conda, but I could not clear package conflict.
So, please help creating baseline image for zorder testing.

I've already removed the broken baseline image from history. (Due to the many conflict when merging, the code is now rebased to current RELEASE_next_minor)

ericpre committed:
Add baseline image for marker plotting test with zorder

ericpre committed:
Improve user guide and docstring: use intersphinx links instead to https links

ericpre commented:
@nem1234, I added a baseline images and improve formatting and links in the user guide and docstring.

ericpre reviewed (commented):
This looks good.
I am wondering if the representation of the markers should be simplified even further by removing the `zorder` information?

nem1234 committed:
Simplify document for marker properties

ericpre commented on hyperspy/drawing/_markers/vertical_line_segment.py:
```diff
@@ -58,20 +58,29 @@ class VerticalLineSegment(MarkerBase):
+            zorder = self.marker_properties['zorder']
+        else:
+            zorder = None
+        string = "<marker.{}, {} (x={},y1={},y2={},color={},zorder={})>".format(
```
Sorry, I should have been more explicit, I meant the `zorder` in the `__repr__` method, as for example here.

nem1234 replied:
This is a debugging code and was used for checking version dependendancy of matplotlib.
Meaning of zorder=None is different for some (not all) markers in matplotlib 3.1.3 and 3.4.3, and the property of zorder=None in kwargs should be removed in 3.1.3 or earlier versions.
Now, I understood how zorder works on matplotlib versions, I can remove zorder in __repr__,
and also I can be remove the zorder initialization code in marker initialization.
I will try to simplify them.

nem1234 committed:
Remove default setting of zorder from drawing._markers.*
Remove zorder from __repr__
Revise test code for zorder in __repr__

ericpre commented on hyperspy/drawing/_markers/horizontal_line_segment.py:
```diff
@@ -64,6 +64,10 @@ def __init__(self, x1, x2, y, **kwargs):
         self.marker_properties = lp
         self.set_data(x1=x1, x2=x2, y1=y)
         self.set_marker_properties(**kwargs)
+        # for compatibility with matplotlib 3.1.x - 3.4.x
+        if 'zorder' in self.marker_properties:
+            if self.marker_properties['zorder'] is None:
+                del self.marker_properties['zorder']
```
This behaviour can be confusing for the user as it would remove the `zorder` without warning or error for this specific marker. Instead, I would suggest the following:
```suggestion
        if (Version(maptlotlib.__version__) <= Version("3.4")
            and "zorder" in kwargs.keys()):
            _logger.warning("`zorder` kwargs is only supported "
                            "with matplotlib >= 3.5.")
        self.set_marker_properties(**kwargs)
```

This would have the advantage of not hiding anything to the user and to allow using "zorder" if matplotlib supports it.

The other alternative is to do nothing and ignore if the `zorder` is passed and matplotlib doesn't support it. In this case, matplotlib should raise an error or a warning - I haven't check if it does so, but this is what I would expect!

nem1234 replied:
The zorder property itself is suppoted on matplotlib 3.1,x, but the behaviour of zorder=None is different for some objects.
I already removed the default value of zorder=None, so, I can remove these lines.
(In such case, we need to write version dependent code on users script like you shows)

ericpre commented on upcoming_changes/2871.enhancements.rst:
```diff
@@ -0,0 +1,4 @@
+* Enhancement of drawing.marker
+  1. Add new marker "arrow" and "ellipse"
+  2. Add support for filled rectangle
+  3. Add "zorder" property for markers
```
This should be changed depending on what is decided with my comment about `zorder` above.

nem1234 replied:
I removed the 3rd line.

ericpre reviewed (commented):
On last comment! :)

nem1234 committed:
Remove codes for zorder property

ericpre reviewed (approved):
This looks good to me. @jlaehne, as you started a review, would you like to do another round?

jlaehne commented on doc/user_guide/visualisation.rst:
```diff
@@ -1332,3 +1331,45 @@ Permanent markers are stored in the HDF5 file if the signal is saved:
+
+
+Markers properties
+------------------
```
'Marker properties'

jlaehne commented on hyperspy/drawing/_markers/arrow.py:
```diff
@@ -0,0 +1,106 @@
+
+    def update(self):
+        if self.auto_update is False:
+            return
```
`auto_update = False` not tested for

nem1234 replied:
Add a test for both auto_update == True and False

jlaehne commented on hyperspy/drawing/_markers/ellipse.py:
```diff
@@ -0,0 +1,105 @@
+
+    def update(self):
+        if self.auto_update is False:
+            return
```
as above

jlaehne commented on hyperspy/drawing/_markers/ellipse.py:
```diff
@@ -0,0 +1,105 @@
+        try:
+            self.marker.set_width(width)
+            self.marker.set_height(height)
+        except: # for matplotlib 3.1.x
```
is it possible to test for 3.1.x specific functionalities?

nem1234 replied:
In my knowledge, "pytest --mpl"  runs under matplotlib 3.1.3 in azure pipeline.
But this code is incorrect and not work correctly.
Fixed.

jlaehne commented on hyperspy/drawing/_markers/point.py:
```diff
@@ -70,10 +69,14 @@ class Point(MarkerBase):
         self.marker_properties = lp
         self.set_data(x1=x, y1=y, size=size)
         self.set_marker_properties(**kwargs)
+        # for compatibility with matplotlib 3.1.x - 3.4.x
```
is it possible to test for 3.1.x specific functionalities?

nem1234 replied:
I'll remove this code. (Default value of zorder=None is already removed.)

jlaehne commented:
OK, I added a few more (minor) comments, would be great if the typos could be corrected and coverage completed.

jlaehne commented on upcoming_changes/2871.enhancements.rst:
```diff
@@ -0,0 +1,3 @@
+* Enhancement of drawing.marker
```
```Suggestion
Enhancement of drawing.marker
```

nem1234 replied:
revised.

jlaehne commented on upcoming_changes/2871.enhancements.rst:
```diff
@@ -0,0 +1,3 @@
+* Enhancement of drawing.marker
+  1. Add new marker "arrow" and "ellipse"
```
```Suggestion
  1. Add new markers "arrow" and "ellipse"
```

Would also make sense to put all on one line.

For the formatting in the changelog, see https://hyperspy--2871.org.readthedocs.build/en/2871/user_guide/changes.html
Currently, there is a double bullet.

nem1234 replied:
Thank you.
revised to 1-line.

nem1234 committed:
Add test code for auto_update/fixed markers with navigation window
Remove zorder test from point marker (not used)
Fix auto_update code in ellipse under matplotlib 3.1.x, 3.2.x
Revise link in visualization.rst

nem1234 committed:
Simplify change log

nem1234 committed:
Revise link to EDS tutorial

jlaehne commented on upcoming_changes/2871.enhancements.rst:
```diff
@@ -0,0 +1 @@
+Add new markers :py:meth:`~.utils.markers.arrow`, :py:meth:`~.utils.markers.ellipse` and filled :py:meth:`~.utils.markers.rectangle`.
```
```suggestion
Add new markers :py:meth:`~drawing._markers.arrow`, :py:meth:`~.drawing._markers.ellipse` and filled :py:meth:`~.drawing._markers.rectangle.
```

jlaehne replied:
Otherwise, the links do not work.

nem1234 committed:
add marker.name for ellipse
simplify test code

nem1234 committed:
add test for horizontal_line_segment() and vertical_line_segment() with parameters of None

nem1234 committed:
Fix array size when testing horizontal_line_segment and vertical_line_segment with parameters of None

nem1234 committed:
remove debug code

jlaehne committed:
Update upcoming_changes/2871.enhancements.rst

jlaehne commented on upcoming_changes/2871.enhancements.rst:
```diff
@@ -1 +1 @@
-Add new markers :py:meth:`~.utils.markers.arrow`, :py:meth:`~.utils.markers.ellipse` and filled :py:meth:`~.utils.markers.rectangle`.
+Add new markers :py:meth:`~drawing._markers.arrow`, :py:meth:`~.drawing._markers.ellipse` and filled :py:meth:`~.drawing._markers.rectangle.
```
```suggestion
Add new markers :py:mod:`~.drawing._markers.arrow`, :py:mod:`~.drawing._markers.ellipse` and filled :py:mod:`~.drawing._markers.rectangle`.
```

jlaehne committed:
Update upcoming_changes/2871.enhancements.rst

magnunor commented:
Nice! I'll have a look at https://github.com/hyperspy/hyperspy/pull/1491 later today.
